app/code/Magento/Catalog/Controller/Adminhtml/Product/Validate.php

```diff
@@ -149,6 +184,45 @@ public function execute()
         return $this->resultJsonFactory->create()->setData($response);
     }
 
+    /**
+     * Validates Url Key uniqueness.
+     *
+     * @param ProductModel $product
+     * @throws UrlAlreadyExistsException
+     */
+    private function validateUrlKeyUniqueness(ProductModel $product): void
```

Comment:
I believe we should create a new class/model for this validation, we may need to use it on another controller. This should be added to CatalogUrlRewrite module and used here, it will reduce complexity on the controller at least.

Reply:
@gabrieldagama Fixed.